qiime2/sdk/result.py

```diff
@@ -282,6 +286,205 @@ def result(self):
+        -----
+            In Archive Format 7.0, `referenced_result_uuid` is set to
+            the same value as `root_result_uuid`, but this will change
+            in future versions to allow for Annotations that may reference
+            a different Result than the one they are attached to.
+
+        See Also
+        --------
+        Annotation.write
+
+        """
+        self._validate_annotation_support()
+        # Guard to ensure Annotation names are unique per Result object
+        for existing_annotation in self._annotations:
+            if annotation.name == existing_annotation.name:
+                raise ValueError(
+                    'Namespace collision occurred when attempting to add '
+                    f'Annotation with name: "{annotation.name}"\n'
+                    'Annotation names must be unique within each Result '
+                    'they are attached to.'
+                )
+
+        annotation.write(annotations_dir=self._archiver.annotations_dir,
+                         root_result_uuid=str(self.uuid),
+                         referenced_result_uuid=str(self.uuid))
+        self._annotations.append(annotation)
+
+        # now calculate checksums for all files within the newly minted
+        # annotation subdir
+        annotation_dir = \
+            pathlib.Path(self._archiver.annotations_dir) / str(annotation.uuid)
+        checksum_ext = self._archiver._fmt.CHECKSUM_TYPE
+        manifest = self._archiver._fmt.CHECKSUM_FILE
+
+        checksums = util.checksum_directory(str(annotation_dir),
+                                            checksum_type=checksum_ext)
+        checksums.pop(manifest, None)
+
+        with (annotation_dir / manifest).open('w') as fh:
+            for item in checksums.items():
+                fh.write(util.to_checksum_format(*item))
+                fh.write('\n')
+
+    def get_annotation(self, name):
+        """Retrieve an Annotation given by `name` from the Result object.
+
+        Parameters
+        ----------
+        name : str\n
+            The name of the Annotation to retrieve.
+
+        Returns
+        -------
+        Annotation : obj\n
+            The Annotation object associated with the provided name.
+
+        Raises
+        ------
+        ValueError\n
+            If no Annotation with the provided name is found.
+
+        """
+        self._validate_annotation_support()
+
+        for annotation in self._annotations:
+            if annotation.name == name:
+                return annotation
+
+        raise ValueError(f'No Annotation with name: "{name}" was found.')
+
+    # TODO: add support to filter by type
+    # once additional annotation types are added in 7.1
+    def iter_annotations(self):
+        """Constructs an iterable containing all Annotations associated with
+        the Result object.
+
+        Raises
+        ------
+        ValueError\n
+            If no Annotations are found associated with the Result.
```

Comment:
This one might be better to not raise an error from, in which case you can implement it as:
```python
yield from self._annotations
```
and that's basically the whole method.